Introduce a strict Content-Security-Policy

ryanpetrello opened the pull request:
see: https://github.com/ansible/awx/issues/2056

ryanpetrello committed:
Introduce a strict Content-Security-Policy

unlikelyzero commented:
Given that the pipes are borked. I think we'll have to run against a yolo'd instance locally. The key areas to test manually would be file upload and download on windows.

jakemcdermott committed:
Integrate CSP config with frontend framework

jakemcdermott committed:
Add w3c-compliant reporting for CSP violations

jakemcdermott committed:
Add system for strict-csp stdout html generation

jakemcdermott commented:
recheck

ryanpetrello reviewed (commented):
I poked at this for awhile, and it seems to work really well for me 🎉!

ryanpetrello reviewed (commented):
@jakemcdermott after this merges, gimme a heads up, because we'll need to remove the CSP header from some downstream nginx configs, too.

ryanpetrello commented:
This PR is *just about* wrapped up, but I think @nixocio is going to add a server-side endpoint for reporting CSP violations (at `/csp-violation/`) so that we can spot violations in our server logs.

Probably the simplest thing to do here is to report this via the general AWX logger e.g.,

`logging.getLogger('awx').error(...)`

ryanpetrello commented on awx/ui_next/public/index.html:
```diff
@@ -1,6 +1,15 @@
 <!DOCTYPE html>
 <html lang="en">
   <head>
+    <% if (process.env.NODE_ENV === 'production') { %>
```
@jakemcdermott how does this get set?  Will this take effect for official AWX builds?

jakemcdermott replied:
Yes

jakemcdermott replied:
> how does this get set?

This isn't the final `index.html` that gets packaged in the build or collected by django. It's actually an `.ejs` template. 

Were it configurable by `react-scripts` (it isn't), I would name this file `index.html.j2.ejs`. `react-scripts` processes it as an `.ejs` template to produce the `jinja` template that django uses to serve the actual html file.

jakemcdermott replied:
> Will this take effect for official AWX builds?

Yes

> how does this get set?

This isn't the final index.html that gets packaged in the build and collected by django. It's actually (and always has been since the angular days) an `.ejs` template.

When you run `make ui-devel`, react-scripts processes it as an `.ejs` template and outputs the jinja/djt template that django uses to render the actual index.html in production. Were it configurable by react-scripts (it isn't), I would rename this source file to `index.html.j2.ejs`. and the one django consumes to `index.html.j2` to make this pipeline more explicit.

nixocio replied:
Good explanation @jakemcdermott

ryanpetrello commented on awx/ui_next/src/setupCSP.js:
```diff
@@ -0,0 +1,30 @@
+
+  const xhr = new XMLHttpRequest();
+  xhr.open('POST', '/csp-violation/', true);
+  xhr.setRequestHeader('content-type', 'application/csp-report');
```
This may need to be `application/json` for DRF to be happy (and not require a bunch of weird gymnastics w/ DRF POST body parsers).

jakemcdermott replied:
The specification is to use a `Content-Type` header field of `application/csp-report`: https://w3c.github.io/webappsec-csp/2/#violation-reports

jakemcdermott replied:
@nixocio Tells me this is already wired up: https://github.com/ansible/awx/blob/devel/awx/urls.py#L25

![Screenshot from 2020-12-07 20-52-47](https://user-images.githubusercontent.com/9753817/101428233-4f501980-38ce-11eb-82f5-a78b72766495.png)

:man_shrugging:

ryanpetrello replied:
https://github.com/ansible/awx/commit/eacf819caf4

<img src="https://images.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia.giphy.com%2Fmedia%2FFPjbHO0jJxGsE%2Fgiphy.gif&f=1" />

ryanpetrello commented:
recheck

ryanpetrello commented:
@nixocio @jakemcdermott I'm still seeing some UI test failures on this PR:

https://ansible.softwarefactory-project.io/logs/54/8754/75db4c9437bf5f4dbf32a5b69d6f15d772d9068a/check/awx-ui/8a02ba8/log-classify.html

Otherwise, I'm ready to merge this whenever this is addressed and when @unlikelyzero gives a 👍

jakemcdermott committed:
Add global mock for webpack csp var

jakemcdermott committed:
Avoid prop reference error in test

keithjgrant commented on awx/ui_next/public/index.html:
```diff
@@ -1,6 +1,15 @@
+        http-equiv="Content-Security-Policy"
+        content="default-src 'self'; connect-src 'self' ws: wss:; style-src 'self' 'nonce-{{ csp_nonce }}' *.pendo.io; script-src 'self' 'nonce-{{ csp_nonce }}' *.pendo.io; img-src 'self' *.pendo.io data:;"
+      />
+    <% } %>
```
Is it okay to use the same nonce for styles and scripts? Since a nonce can "only be used once," I'm vaguely worried some browser may reject the second one

jakemcdermott replied:
That's a good question. I _think_ we're ok there. 

My understanding is that the nonce should be used only once _per page load_. The way `__webpack_nonce__` works seems to support this, in that it takes one `nonce` value and applies it globally to both `style-src` and `inline-src`.

ryanpetrello replied:
Yep, the nonce is intended as a per-page load nonce, I think (what @jakemcdermott said).  I don't think two separate nonces buys us anything - it's just two server-side secrets we'd end up generating per request instead.

Maybe there's some browser out there that treats this differently, but my Chrome and Safari testing seems to work fine.  If anybody's breaking the rules or has weird support, you know which vendor it's probably gonna be 😉.

nixocio replied:
~I used firefox with those changes and no errors reported~.  Well, errors were reported just down below.

jakemcdermott committed:
Move INLINE_RUNTIME_CHUNK flag to scripts definition

We _always_ want INLINE_RUNTIME_CHUNK to be false when building the ui,
even if someone happens to unexpectedly make a production build without
using the top-level make targets for some reason.

nixocio committed:
Fix JobEvent tests

Fix JobEvent tests

unlikelyzero commented:
Tested it based on a build from 9am. Looks like there's an issue with websockets. I need to drop for the next couple hours but I'll retest with a latest build and compare it to devel (just to make sure it's not an existing problem)

unlikelyzero commented:
Does not work with Firefox ESR. Will need to rethink implementation

unlikelyzero reviewed (changes requested):
Seeing these errors in Firefox ESR in Windows 10

Content Security Policy: The page's settings blocked the loading of a resource at inline ("style-src").

jakemcdermott commented:
Confirming what @unlikelyzero is seeing - On Firefox ESR 78:

![Screenshot from 2020-12-09 12-33-40](https://user-images.githubusercontent.com/9753817/101665383-eaeea080-3a1a-11eb-9054-427c7c6cffb3.png)

ryanpetrello commented:
Good find @unlikelyzero @jakemcdermott 😄

jakemcdermott commented:
@unlikelyzero found it! Thank you @unlikelyzero

jakemcdermott committed:
FFox ESR 78 Compatibility

unlikelyzero reviewed (approved):
Follow up ticket for Firefox ESR support filed

swathit0717 commented:
I am running AWX - 16.0.0. Looks like "Introduce a strict Content-Security-Policy #8754" has already been merged to 16.0.0. When I tried to check the CSP header using "CSP Evaluator" Chrome extension, I observe that "No Content Security Policy has been set".
Later, I tried with AWX -15.0.0, there is CSP header already set. Is there a way to modify CSP header to remove the 'unsafe-inline' directive. Please suggest. 
<img width="1611" alt="image" src="https://user-images.githubusercontent.com/81205977/148282415-f19d54b4-8437-47f0-9bed-080647d804e6.png">x
